### ISSUE TITLE
Derivative function fails between new data insertion and wal flush

### ISSUE THREAD

**Issue by gotyaoi** [influxdb/influxdb#4337]
I just upgraded to influxdb 0.9.4.2 and am noticing some very odd behavior. Can't say if this is new in the point release or I just didn't notice it before, but basically, some of my grafana graphs were intermittently empty. I've reproduced this in as minimal a setup as possible, as follows:

centos 6.7
influxdb 0.9.4.2, default config that came with the rpm
telegraf 0.1.9, running the disk plugin every 10 seconds.

Using the influx client, I can watch the data accumulate and make most queries as normal. However, the following query:
```
[query] 2015/10/05 15:48:12 SELECT derivative(mean(value), 1s) FROM "telegraf"."default".io_reads WHERE host =~ /cent-test/ AND time > now() - 5m GROUP BY time(10s)
```
sometimes returns the proper data:
```
> select derivative(mean(value), 1s) from "net_bytes_recv" where time > now() - 5m group by time(10s)
name: net_bytes_recv
--------------------
time                    derivative
2015-10-05T22:32:20Z    2921
2015-10-05T22:32:30Z    1871
2015-10-05T22:32:40Z    2961
2015-10-05T22:32:50Z    2617
2015-10-05T22:33:00Z    3174
2015-10-05T22:33:10Z    9775
2015-10-05T22:33:20Z    15424
2015-10-05T22:33:30Z    1639
2015-10-05T22:33:40Z    2786
2015-10-05T22:33:50Z    3988
2015-10-05T22:34:00Z    6955
2015-10-05T22:34:10Z    16843
2015-10-05T22:34:20Z    3768
2015-10-05T22:34:30Z    1487
2015-10-05T22:34:40Z    2162
2015-10-05T22:34:50Z    2284
2015-10-05T22:35:00Z    1714
2015-10-05T22:35:10Z    1607
2015-10-05T22:35:20Z    2764
2015-10-05T22:35:30Z    6220
2015-10-05T22:35:40Z    1789
2015-10-05T22:35:50Z    5590
2015-10-05T22:36:00Z    1505
2015-10-05T22:36:10Z    243
2015-10-05T22:36:20Z    1400
2015-10-05T22:36:30Z    2290
2015-10-05T22:36:40Z    2910
2015-10-05T22:36:50Z    3123
2015-10-05T22:37:00Z    6208
```
and sometimes returns nothing:
```
> select derivative(mean(value), 1s) from "net_bytes_recv" where time > now() - 5m group by time(10s)
name: net_bytes_recv
--------------------
time                    derivative
2015-10-05T22:29:00Z    0
```

Watching the logs, I've determined the difference to be that the query fails in the time between the insertion of the data:
```
[http] 2015/10/05 15:48:05 ::1 - - [05/Oct/2015:15:48:05 -0700] POST /write?consistency=&db=telegraf&precision=s&rp= HTTP/1.1 204 0 - InfluxDBClient 2462683f-6bb3-11e5-827c-000000000000 24.152065ms
```
and the wal flush:
```
[wal] 2015/10/05 15:48:11 Flush due to idle. Flushing 102 series with 102 points and 1733 bytes from partition 1
[wal] 2015/10/05 15:48:11 write to index of partition 1 took 91.148896ms
```
After the flush and before the next insertion, the query is successful.

**Comment by sebito91** [comment]
Add fill(none) to the end of your query!

On Monday, October 5, 2015, gotyaoi <notifications@github.com> wrote:

> I just upgraded to influxdb 0.9.4.2 and am noticing some very odd
> behavior. Can't say if this is new in the point release or I just didn't
> notice it before, but basically, some of my grafana graphs were
> intermittently empty. I've reproduced this in as minimal a setup as
> possible, as follows:
>
> centos 6.7
> influxdb 0.9.4.2, default config that came with the rpm
> telegraf 0.1.9, running the disk plugin every 10 seconds.
>
> Using the influx client, I can watch the data accumulate and make most
> queries as normal. However, the following query:
>
> [query] 2015/10/05 15:48:12 SELECT derivative(mean(value), 1s) FROM "telegraf"."default".io_reads WHERE host =~ /cent-test/ AND time > now() - 5m GROUP BY time(10s)
>
> sometimes returns the proper data:
>
> > select derivative(mean(value), 1s) from "net_bytes_recv" where time > now() - 5m group by time(10s)
> name: net_bytes_recv
> --------------------
> time                    derivative
> 2015-10-05T22:32:20Z    2921
> 2015-10-05T22:32:30Z    1871
> 2015-10-05T22:32:40Z    2961
> 2015-10-05T22:32:50Z    2617
> 2015-10-05T22:33:00Z    3174
> 2015-10-05T22:33:10Z    9775
> 2015-10-05T22:33:20Z    15424
> 2015-10-05T22:33:30Z    1639
> 2015-10-05T22:33:40Z    2786
> 2015-10-05T22:33:50Z    3988
> 2015-10-05T22:34:00Z    6955
> 2015-10-05T22:34:10Z    16843
> 2015-10-05T22:34:20Z    3768
> 2015-10-05T22:34:30Z    1487
> 2015-10-05T22:34:40Z    2162
> 2015-10-05T22:34:50Z    2284
> 2015-10-05T22:35:00Z    1714
> 2015-10-05T22:35:10Z    1607
> 2015-10-05T22:35:20Z    2764
> 2015-10-05T22:35:30Z    6220
> 2015-10-05T22:35:40Z    1789
> 2015-10-05T22:35:50Z    5590
> 2015-10-05T22:36:00Z    1505
> 2015-10-05T22:36:10Z    243
> 2015-10-05T22:36:20Z    1400
> 2015-10-05T22:36:30Z    2290
> 2015-10-05T22:36:40Z    2910
> 2015-10-05T22:36:50Z    3123
> 2015-10-05T22:37:00Z    6208
>
> and sometimes returns nothing:
>
> > select derivative(mean(value), 1s) from "net_bytes_recv" where time > now() - 5m group by time(10s)
> name: net_bytes_recv
> --------------------
> time                    derivative
> 2015-10-05T22:29:00Z    0
>
> Watching the logs, I've determined the difference to be that the query
> fails in the time between the insertion of the data:
>
> [http] 2015/10/05 15:48:05 ::1 - - [05/Oct/2015:15:48:05 -0700] POST /write?consistency=&db=telegraf&precision=s&rp= HTTP/1.1 204 0 - InfluxDBClient 2462683f-6bb3-11e5-827c-000000000000 24.152065ms
>
> and the wal flush:
>
> [wal] 2015/10/05 15:48:11 Flush due to idle. Flushing 102 series with 102 points and 1733 bytes from partition 1
> [wal] 2015/10/05 15:48:11 write to index of partition 1 took 91.148896ms
>
> After the flush and before the next insertion, the query is successful.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/influxdb/influxdb/issues/4337>.
>


-- 

Sebastian Borza
---------------
PGP: EDC2 BF61 4B91 14F2 AAB4  06C9 3744 7F3F E411 0D3E

**Comment by gotyaoi** [comment]
Ok, that seems to be working. Now the question becomes why does it work and where is it documented?

**Comment by jwilder** [comment]
Should be fixed via #4292 which is in the current nightly.

**Comment by gotyaoi** [comment]
Excellent, thank you.
